add `OpInfo` for `torch.nn.functional.kl_div`

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65681
* #68419
* #69212
* #67079
* #67023
* #67022
* #69211
* #65680
* __->__ #65469

Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc @albanD @mruberry @jbschlosser @walterddr @ezyang @gchanan

pmeier committed:
add `OpInfo` for `torch.nn.functional.kl_div`

[ghstack-poisoned]

pmeier commented:
Failures are expected. PR is blocked until #65466 is resolved.

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

[ghstack-poisoned]

pmeier commented:
The latest commit cleans up the implementation of `kl_div` by quite a bit, which ultimately also resolved #65466:

- remove the split between `kl_div_backward_cpu` and `kl_div_backward_cuda`
- remove the explicit double backward

pmeier commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1163,28 +1163,30 @@ Tensor infinitely_differentiable_logit_backward(
-Tensor kl_div_target_backward(Tensor grad_output, Tensor self, Tensor target, int64_t reduction, bool log_target) {
-  Tensor grad_target;
-  if (!log_target) {
-    grad_target = grad_output.mul(target.log().add_(1).sub_(self)).masked_fill_(target == 0, 0.);
```
The `masked_fill` is actually wrong here. While `x * log(x)` for `x == 0` is commonly defined as `0`, its gradient `1 + log(x)` is `-inf` there.

Lezcano replied:
Nit. It's not "commonly defined as 0, but it so happens that lim x -> 0 xlog(x) = 0.

Lezcano commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1163,28 +1163,30 @@ Tensor infinitely_differentiable_logit_backward(
-Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, bool log_target) {
-  auto result = kl_div_backward(grad, input, target, at::Reduction::None, log_target);
+Tensor kl_div_backward(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
+  auto grad_expand = grad.expand_as(input);
```
You can probably delete this line.

Lezcano commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1163,28 +1163,30 @@ Tensor infinitely_differentiable_logit_backward(
-    grad_target = grad_output.mul(target.add(1).sub_(self).mul_(target.exp()));
-  }
+Tensor kl_div_target_backward(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
+  auto grad_expand = grad.expand_as(input);
```
Same as above, you can probably delete this line.

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5672,6 +5672,29 @@ def sample_inputs_pixel_unshuffle(op_info, device, dtype, requires_grad, **kwarg
+        ((2,), dict()),
+        ((2, 3), dict()),
+        ((2, 3, 4), dict()),
+        ((2,), dict(log_target=True)),
+        ((2,), dict(reduction="none")),
+        ((2,), dict(reduction="batchmean")),
+        ((2,), dict(reduction="sum")),
```
Given that the `log_target` takes different routes in the backward, might be good to test the 8 combination log_target x reduction?

Lezcano commented on aten/src/ATen/core/Reduction.h:
```diff
@@ -10,6 +10,7 @@ enum Reduction {
   None,             // Do not reduce
   Mean,             // (Possibly weighted) mean of losses
   Sum,              // Sum losses
+  BatchMean,        // Average only over batch dimension
```
Is it fine to add this here without it being supported by any of the other losses that accept a reduction? What happens if we pass `"batchmean"` as the reduction to any of the other losses?

pmeier replied:
You are right, that shouldn't be there. It is handled completely on the Python side:

https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/nn/functional.py#L2801-L2803

I wonder where the name `"batchmean"` came from if its apparently the same as `"sum"`.

lezcano reviewed (commented):
This looks good, I just left a few comments.

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_nn.py:
```diff
@@ -513,7 +513,7 @@ def kldivloss_with_target_no_reduce_test():
 
 def kldivloss_no_reduce_test():
-    t = torch.randn(10, 10)
+    t = torch.rand(10, 10)
```
The target of `KLDivLoss` can either be a probability `[0, 1]` or a log-probability `(-inf, 0]`. I've fixed the sample inputs here, but there are some more failures due to 

https://github.com/pytorch/pytorch/blob/facff2ec6562d61eab2ccd9f1c16872e54eec92d/torch/testing/_internal/common_nn.py#L5709

in tests that are automatically generated. Not sure what to do about them.

pmeier replied:
The failures are semi-legit:

- they actually show a divergence between the forward and backward pass:
- the divergence is caused by a really odd choice in the forward pass https://github.com/pytorch/pytorch/blob/6a990535154116960767ad422e724250ecfb8325/aten/src/ATen/native/Loss.cpp#L114-L117
  This silently ignores negative inputs in case of a non-log target. Not only is this not documented, but also indicates a bug in the user pipeline. 

Although this would be BC breaking, IMO we should remove this special case for negative values. Of course we still need to keep the continuous extension for `0 * log(0) := 0` in the forward pass.

Lezcano replied:
So, the KL divergence is just defined for for probabilities in [0,1]. Even more, there's nothing stopping you from defining the KL-divergence for general measures, so you could think that the inputs are in [0, infty]. Now, enforcing this would mean checking every input which requires some extra computations.

In my opinion, I think it is good to not check that the inputs are positive, and if they are not, you'll end up returning a `-inf`. If this is well-documented and tested, I think it's better than an arbitrary choice of `0` that's not documented.

Now, this is BC-breaking, but it's BC-breaking with respect to an undocumented "feature", so I believe we should be free to fix these things. cc @jbschlosser

ezyang commented:
Backwards removal is technically BC breaking, marking as such. (I'm not too sure if it should be removed here or not, hopefully someone who is more familiar here can make the call.)

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_nn.py:
```diff
@@ -4666,18 +4666,19 @@ def padding3d_circular(input, pad):
         module_name='KLDivLoss',
-        input_fn=lambda: torch.rand(10, 10).log(),
-        target_fn=lambda: torch.rand(10, 10),
+        constructor=wraps(nn.KLDivLoss)(partial(nn.KLDivLoss, log_target=True)),
```
This is an ugly hack, but I feel like this whole file is. Basically, this is a one-liner that is able to create `nn.KLDivLoss` with `log_target=True` and also have `.__name__ == "KLDivLoss"`. cc @thomasjpfan @jbschlosser

pmeier commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5471,7 +5473,7 @@ def flatten(xs):
 # Check that regression criterion work with no batch dimensions
 regression_criterion_no_batch = [
-    'L1Loss', 'MSELoss', 'PoissonNLLLoss', 'KLDivLoss', 'HuberLoss', 'SmoothL1Loss'
+    'L1Loss', 'MSELoss', 'PoissonNLLLoss', 'HuberLoss', 'SmoothL1Loss'
```
We need to exclude `KLDivLoss` here, because just setting the input and target sizes (L5483-L5484) will create the inputs with `randn` (see https://github.com/pytorch/pytorch/pull/65469/files#r716708268). This would result in NaN's which would fail the gradcheck.

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

lezcano reviewed (approved):
Just left a couple small comments. Everything else looks good modulo making the CI go green.

cc @JackCaoG regarding the XLA failure due to moving `kl_backward` to `FunctionsManual.cpp`

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier commented:
@JackCaoG What is the status on the XLA side?

JackCaoG commented:
@pmeier working on https://github.com/pytorch/xla/pull/3185, fixing one last test. Should be good to go pretty soon.

JackCaoG commented:
@pmeier xla pr is ready

pmeier commented:
Ping @mruberry @jbschlosser for review.

mruberry commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -106,46 +106,10 @@ Tensor margin_ranking_loss(const Tensor& input1, const Tensor& input2, const Ten
-  }
-  return grad_input;
+  auto output = log_target ? at::exp(target) * (target - input)
+                           : at::xlogy(target, target) - target * input;
```
This is technically a different function than what was previously implemented, right? Would you elaborate on the change here?

Lezcano replied:
It is indeed different in that the KL of a negative number this implementation returns `NaN`, while before it returned zero. This is consistent with the formula. Even more, it is consistent with the idea that the function is not defined for negative numbers.
On a side note, it's also faster to do this than to return zeros.

Lezcano replied:
This is also consistent with the current state of the docs added here: https://github.com/pytorch/pytorch/pull/67443

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2383,11 +2383,6 @@
   dispatch:
     CompositeExplicitAutograd: kl_div
 
-- func: kl_div_backward(Tensor grad_output, Tensor self, Tensor target, int reduction=Mean, *, bool log_target=False) -> Tensor
```
I don't think anyone is using this but this is, of course, a BC-breaking change.

mruberry commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -7044,16 +7044,16 @@ def forward(self, input, target):
 
-        x = torch.randn(5)
-        y = torch.randn(5)
+        x = torch.rand(5).log()
```
Nice ONNX fixes

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1266,30 +1266,25 @@ Tensor infinitely_differentiable_logit_backward(
-    grad_target = grad_output.mul(target.add(1).sub_(self).mul_(target.exp()));
-  }
-
+Tensor kl_div_target_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
```
Would you elaborate on the changes here, too?

Lezcano replied:
Same thing as in the other case, it concerns negative numbers, for which this function is not defined. It is also faster to do it this way, and it's also consistent with the docs.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1266,30 +1266,25 @@ Tensor infinitely_differentiable_logit_backward(
 
-Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, bool log_target) {
-  auto result = kl_div_backward(grad, input, target, at::Reduction::None, log_target);
+Tensor kl_div_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
```
Isn't this going to be a performance regression over the previous implementation because previously there were handwritten fused kernels for these operations?

Lezcano replied:
Potentially, although just for CUDA and for `log_target=False` as the TI was just on that branch. It should be faster on the CPU side though.

I would not worry too much about this, as, in that case, just 3 kernels are being called. I reckon that this is not such a big deal (?)

mruberry reviewed (commented):
Wow there is a lot going on here!

Let's separate the OpInfo, which looks great, from the other changes. For the other changes I have some concerns that I added inline comments for. Looking forward to discussing them!

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

pmeier commented:
After some offline discussion with @mruberry and @Lezcano, we opted to separate the BC breaking part into a another PR (#68419), but keep everything else as is.

Lezcano commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -46,32 +46,6 @@ void binary_cross_entropy_backward_out_kernel(Tensor& grad_input, const Tensor&
 
 namespace at { namespace native {
 
-Tensor kl_div_backward_cuda(const Tensor& grad, const Tensor& input, const Tensor& target, int64_t reduction, bool log_target) {
```
These function and the `_cpu` counterpart you should keep and remove in the other PR.

Lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2386,8 +2386,7 @@
 
 - func: kl_div_backward(Tensor grad_output, Tensor self, Tensor target, int reduction=Mean, *, bool log_target=False) -> Tensor
   dispatch:
-    CPU: kl_div_backward_cpu
-    CUDA: kl_div_backward_cuda
```
Should leave as it was and just remove the whole entry in the next PR.

Lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -1975,11 +1975,6 @@
   grad_output: hardtanh_backward(grad, self, min_val, max_val)
   self: zeros_like(grad)
 
-- name: kl_div_backward(Tensor grad_output, Tensor self, Tensor target, int reduction=Mean, *, bool log_target=False) -> Tensor
```
In `derivatives.yaml` you'll need to change the derivative of `kl_div` to be `kl_div_backward_aux`, and then rename it to `kl_div_backward` in the next PR.

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc ezyang gchanan mruberry jbschlosser walterddr

[ghstack-poisoned]

lezcano reviewed (approved):
LGTM if the CI agrees!

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all -l ciflow/slow

mruberry commented:
Unfortunately @jbschlosser is under the weather at the moment but he should be better soon and we'll take a look

mruberry reviewed (commented):
Hey @pmeier, @Lezcano! Thanks for your patience.

@mruberry and @jbschlosser here (@jbschlosser is feeling much better!). Here's our take on this PR:

- it's cool, but...
- we don't want to change the semantics of kl_div in this PR
- would you open an issue so we can discuss the NaN semantics more? 
- the test changes to ensure the inputs to kl_div are valid are awesome
- the OpInfo is cool, of course

So basically we'd like to shrink the PR's semantic changes and open an issue to discuss them separately. What are your thoughts?

pmeier commented:
> So basically we'd like to shrink the PR's semantic changes and open an issue to discuss them separately. What are your thoughts?

This PR adds two changes to `kl_div`:

1. Entangles the logic, which should result in a speed-up since we are calling fewer kernels.
2. Removes undocumented behavior, which silently accept wrong inputs. Without `log_target=True`, `target` should be a probability in `[0.0, 1.0]`. The current implementation simply substitutes `0` for the loss in case it encounters a negative `target`:

    https://github.com/pytorch/pytorch/blob/fbaa19a6fa6c7d07ecd74e5ebf3c5d0fc77ecc51/aten/src/ATen/native/Loss.cpp#L117

    For example:
    
    ```python
    import torch
    from torch.nn.functional import kl_div
    
    input = torch.rand(()).log()
    target = torch.tensor(-1, dtype=input.dtype)
    
    print(kl_div(input, target))
    ```
    ```
    tensor(0.)
    ```

    This is problematic for two reasons:

    1. It is mathematically wrong, since the equation (in particular `log(target)`) is undefined for negative targets.
    2. It most likely hides a bug in the user code.

    Running the snippet above with the fixes in this PR gives `tensor(nan)` which resolves both issues.

Do you really want me to revert 2. to the old behavior?

Lezcano commented:
I think what @pmeier mentioned is on point. The new implementation is faster and correct (follows the new docs we wrote for this op).
This looks like a similar case to that of `nll_loss`, only that in this case it is clear what the value should be. If the user passes a negative value to this function with `log_target=True` then that should return `NaN` by definition. 
cc @ngimel to chip in as well.

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr ezyang gchanan

[ghstack-poisoned]

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr ezyang gchanan

[ghstack-poisoned]

pmeier commented:
After some more offline discussion, we decided to split all changes to the op into a different PR #69212. Thus, this PR no longer fixes #65466.

saketh-are commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7322,6 +7322,29 @@ def sample_inputs_allclose(op_info, device, dtype, requires_grad, **kwargs):
+        input = make(shape, low=None, high=0)
+        target = make(shape, low=None, high=0) if log_target else make(shape, low=0, high=1)
+        sample_inputs.append(
+            SampleInput(input, args=(target,), kwargs=dict(reduction=reduction, log_target=log_target))
```
Should we consider adding samples which pass `size_average` or `reduce` kwargs for completeness?

pmeier replied:
I'd rather not. They are deprecated since `torch==0.4.1`, which was released over 2 years ago:

https://github.com/pytorch/pytorch/blob/a24163a95edb193ff7b06e98cd69bf7cfd4c0d2f/torch/nn/functional.py#L38

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr ezyang gchanan

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7500,6 +7500,29 @@ def sample_inputs_allclose(op_info, device, dtype, requires_grad, **kwargs):
+    sample_inputs = []
+    for (shape, reduction), log_target in itertools.product(shapes_and_reduction, (True, False)):
+        input = make(shape, low=None, high=0)
+        target = make(shape, low=None, high=0) if log_target else make(shape, low=0, high=1)
```
Let's add a comment explaining log target vs not and why we want the range we do for each case

mruberry reviewed (approved):
Cool! -- would you just add one comment and ping me to merge this, @pmeier?

pmeier committed:
Update on "add `OpInfo` for `torch.nn.functional.kl_div`"


Differential Revision: [D31111698](https://our.internmc.facebook.com/intern/diff/D31111698)

cc albanD mruberry jbschlosser walterddr ezyang gchanan

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31111698).
